tests/integration/TestTransaction.cc

```diff
@@ -0,0 +1,322 @@
+
+        const Topic     topic     = Utility::getRandomString();
+        KafkaTestUtility::CreateKafkaTopic(topic, 1, 3);
+
+        sendMessageWithTransactions("message to commitTransaction", topic, TransactionClosureAction::ToComplete);
+
+        const auto messages = receiveMessages(topic, IsolationLevel::ReadCommitted);
+        ASSERT_EQ(1, messages.size());
+    }
+
+    {
+        std::cout << "========== Producer: commitTransaction, Consumer: isolation.level=read_uncommitted ==========" << std::endl;
+
+        const Topic     topic     = Utility::getRandomString();
+        KafkaTestUtility::CreateKafkaTopic(topic, 1, 3);
+
+        sendMessageWithTransactions("message to commitTransaction", topic, TransactionClosureAction::ToComplete);
+
+        const auto messages = receiveMessages(topic, IsolationLevel::ReadUnCommitted);
+        ASSERT_EQ(1, messages.size());
+    }
+
+    {
+        std::cout << "========== Producer: abortTransaction, Consumer: isolation.level=read_committed ==========" << std::endl;
+
+        const Topic     topic     = Utility::getRandomString();
+        KafkaTestUtility::CreateKafkaTopic(topic, 1, 3);
+
+        sendMessageWithTransactions("message to abortTransaction", topic, TransactionClosureAction::ToAbort);
+
+        const auto messages = receiveMessages(topic, IsolationLevel::ReadCommitted);
+        ASSERT_EQ(0, messages.size());
+    }
+
+    {
+        std::cout << "========== Producer: abortTransaction, Consumer: isolation.level=unread_committed ==========" << std::endl;
+
+        const Topic     topic     = Utility::getRandomString();
+        KafkaTestUtility::CreateKafkaTopic(topic, 1, 3);
+
+        sendMessageWithTransactions("message to abortTransaction", topic, TransactionClosureAction::ToAbort);
+
+        const auto messages = receiveMessages(topic, IsolationLevel::ReadUnCommitted);
+        ASSERT_EQ(0, messages.size());
+    }
+
+    {
+        std::cout << "========== Producer: no commit/abortTransaction, Consumer: isolation.level=read_committed ==========" << std::endl;
+
+        const Topic     topic     = Utility::getRandomString();
+        KafkaTestUtility::CreateKafkaTopic(topic, 1, 3);
+
+        sendMessageWithTransactions("message with no commit/abortTransaction", topic, TransactionClosureAction::ToAbort);
+
+        const auto messages = receiveMessages(topic, IsolationLevel::ReadCommitted);
+        ASSERT_EQ(0, messages.size());
+    }
+
+    {
+        std::cout << "========== Producer: no commti/abortTransaction, Consumer: isolation.level=unread_committed ==========" << std::endl;
+
+        const Topic     topic     = Utility::getRandomString();
+        KafkaTestUtility::CreateKafkaTopic(topic, 1, 3);
+
+        sendMessageWithTransactions("message with no commit/abortTransaction", topic, TransactionClosureAction::ToAbort);
+
+        const auto messages = receiveMessages(topic, IsolationLevel::ReadUnCommitted);
```

Comment:
Is it possible we may read 1 message?

Reply:
Since the transaction has been aborted, no message could be got; otherwise, if the transaction failed to commit, there might be 1 message (or not).

Reply:
I assume the last two sub-tests were to test NoAction+ReadCommitted and NoAction+ReadUnCommitted